Dapr workshop

Dzvezdana opened the pull request:
Includes:

- Beginner friendly guided workshop and solutions.
- Self-guided advanced examples.
- Updated README with instructions.
- Added empty pdf as placeholder for the slides.

gCaglia commented on workshop/introductory_workshop/03_creating_an_agent.ipynb:
```diff
@@ -0,0 +1,543 @@
+   },
+   "outputs": [],
+   "source": [
+    "react_agent = ReActAgent(\n",
```
I had two issues with this class when using it:
- The `verbose` parameter does not seem to have an impact (on Python 3.12).
- Trying to call the `ReactAgent` constructor a second time raises an error until the NB is restarted.

Both of them are a bit odd, and could also be related to the underlying packages. Maybe I also made a mistake, but either way, it wouldn't be a huge deal for me.

Dzvezdana replied:
I just tried and couldn't reproduce this in Jupyter notebook. I'll keep it in mind it can pop up and ask participants to restart the kernal.

gCaglia commented on workshop/introductory_workshop/03_creating_an_agent.ipynb:
```diff
@@ -0,0 +1,543 @@
+    }
+   ],
+   "source": [
+    "if __name__ == \"__main__\":\n",
```
Maybe we can also do something like this instead:

```python
react_agent = ...

def chat(prompt: str):
    result = asyncio.run(react_agent.run(prompt))
    print(result)
```

gCaglia commented on requirements.txt:
```diff
@@ -0,0 +1,156 @@
+aiohappyeyeballs==2.6.1
```
I had to install prompty separately manually.

gCaglia reviewed (commented):
Looks great, and really nice idea to work with News!

So far, I looked only at the introductory part. I will check out the advanced part tomorrow.

gCaglia commented:
So I just finished reviewing the advanced part as well and I have some feedback points:

Exercise 5 is really cool! And I can see there is a lot going on under the hood. While I think 1-3 can be nice introductions, I would hate for people to miss out on number 5. Maybe 2,3,5 could make sense as an order? Either way though, I think it could be fun to have people get their hands a bit dirtier. Maybe we could remove one of the characters, and just provide a template to the participants with which they can add their own character to the conversation?

Also, not sure on what you were planning in that regard still, but offering some guidance on the difference bewteen the ReAct agent from 3 vs the AssistantAgent from 5 would probably be nice.

So far, I was unable to fully run 6, due to an import error and issues with redis and dapr sidecar, but especially the later two could be because I am using podman on my personal laptop instead of docker. 

In any case, if you would like to discuss anything or if I can help, let me know :)

Btw, I was thinking that I could add an note on the meetup page that a container runtime will be recommended as a prerequisite.

Dzvezdana commented on solutions/introductory_workshop/01_ask_llm.ipynb:
```diff
@@ -0,0 +1,503 @@
+    "response = llm.generate(\n",
+    "    messages=[UserMessage(\"Give me a list of 5 random hobbies. Return just the hobby names.\")]\n",
+    ")\n",
+    "print(\"\\n\")\n",
```
For me it merged them all in one line when using Jupyter notebook, so I added a new like for clarity. I can split them to separate cells alternatively, so they are formatted nicely.

Dzvezdana commented on solutions/introductory_workshop/01_ask_llm.ipynb:
```diff
@@ -0,0 +1,503 @@
+   "source": [
+    "#### Troubleshooting\n",
+    "\n",
+    "1. **API Key Issues**: If you see an authentication error, verify your OpenAI API key in the `.env` file\n",
```
Nice catch! I tried .env initially, but it gets ignored by Jupyter notebook, so I had to do the hackhy thing with the helper Python file. I'll update the docs.

Dzvezdana commented:
> So I just finished reviewing the advanced part as well and I have some feedback points:
> 
> Exercise 5 is really cool! And I can see there is a lot going on under the hood. While I think 1-3 can be nice introductions, I would hate for people to miss out on number 5. Maybe 2,3,5 could make sense as an order? Either way though, I think it could be fun to have people get their hands a bit dirtier. Maybe we could remove one of the characters, and just provide a template to the participants with which they can add their own character to the conversation?
> 
> Also, not sure on what you were planning in that regard still, but offering some guidance on the difference bewteen the ReAct agent from 3 vs the AssistantAgent from 5 would probably be nice.
> 
> So far, I was unable to fully run 6, due to an import error and issues with redis and dapr sidecar, but especially the later two could be because I am using podman on my personal laptop instead of docker.
> 
> In any case, if you would like to discuss anything or if I can help, let me know :)
> 
> Btw, I was thinking that I could add an note on the meetup page that a container runtime will be recommended as a prerequisite.

I like the idea, I was just worried the local setup of Dapr/Docker might take long due to different OS. I changed the order of the exercises so exercise 5 is 4 now :) I also added some info on AssistantAgent and a little exercise where the users can add extra characters as services including a solution.

How about I shorten exercise 1 significantly so we potentially have time for exercise 4 at the end? And maybe indeed we can ask users to install Docker/Podman before the workshop on Meetup?

gCaglia commented:
It's absolutely a valid concern that installation of Dapr and docker could be problematic or take long, and if you don't feel comfortable putting more emphasis on those, I will totally understand.
I just think that dapr is a really exciting piece of software to work with so if we can finish of with a section that's really _cool_, I think people will appreciate that :) 

I will post a note for attendees, recommending to look into getting a container runtime beforehand.

gCaglia commented on README.md:
````diff
@@ -1,50 +1,106 @@
+
+If you are using Jupyter notebook locally, please update `helper_secrets.py` file in `workshop/introductory_workshop`:
+
+```env
````
There is now an env block inside a bash block :)

gCaglia commented:
Also, btw at the moment the PDF is still ToDo, right? No rush about it, just wanting to make sure we don't forget about it.

gCaglia reviewed (approved):
Just went through the slides and I would have two feedback points:

1. On slide 10, the "Final answer" has the same symbol as "Tool A" and "Tool B", fine if on purpose, but I was thinking a speech bubble or something like that will make the point of the graphic more immediately clear :) 
2. I don't see any reference to the more advanced exercises in the slides. Probably you wanted to mention it verbally, but I think for the people that get to it, it will be really helpful to have a rundown on what the different things in the advanced exercise folder are. I see service , workflows, etc...  are explained a bit in the Readme, but to have a visualization of how these entities interact and what their relationship to each other is, can be valuable for people to see.

I will be merging the PR for now, if you would like to act on the above feedback, feel free to make a new one :)
